drivers/i2s/i2s_litex.c

```diff
@@ -0,0 +1,617 @@
+}
+
+/**
+ * @brief Disable i2s device
+ *
+ * @param reg base register of device
+ *
+ * @return N/A
+ */
+static void i2s_disable(int reg)
+{
+	uint8_t reg_data = litex_read8(reg + I2S_CONTROL_REG_OFFSET);
+
+	litex_write8(reg_data & ~(I2S_ENABLE), reg + I2S_CONTROL_REG_OFFSET);
+}
+
+/**
+ * @brief Reset i2s fifo 
+ *
+ * @param reg base register of device 
+ *
+ * @return N/A
+ */
+static void i2s_reset_fifo(int reg)
+{
+	uint8_t reg_data = litex_read8(reg + I2S_CONTROL_REG_OFFSET);
+
+	litex_write8(reg_data | I2S_FIFO_RESET, reg + I2S_CONTROL_REG_OFFSET);
+}
+
+/**
+ * @brief Get i2s format handled by device 
+ *
+ * @param reg base register of device 
+ *
+ * @return currently supported format
+ */
+static int i2s_get_foramt(int reg)
+{
+	uint8_t reg_data = litex_read32(reg + I2S_CONFIG_REG_OFFSET);
+	reg_data &= I2S_CONF_FORMAT_MASK;
+	if (reg_data == 1) {
+		return I2S_FMT_DATA_FORMAT_I2S;
+	} else if (reg_data == 2) {
+		return I2S_FMT_DATA_FORMAT_LEFT_JUSTIFIED;
+	}
+	return -EINVAL;
+}
+
+/**
+ * @brief Get i2s sample width handled by device
+ *
+ * @param reg base register of device 
+ *
+ * @return i2s sample width in bits
+ */
+static uint32_t i2s_get_sample_width(int reg)
+{
+	uint32_t reg_data = litex_read32(reg + I2S_CONFIG_REG_OFFSET);
+	reg_data &= I2S_CONF_SAMPLE_WIDTH_MASK;
+	return reg_data >> I2S_CONF_SAMPLE_WIDTH_OFFSET;
+}
+
+/**
+ * @brief Get i2s audio sampling rate handled by device
+ *
+ * @param reg base register of device 
+ *
+ * @return audio sampling rate in Hz
+ */
+static uint32_t i2s_get_audio_freq(int reg)
+{
+	uint32_t reg_data = litex_read32(reg + I2S_CONFIG_REG_OFFSET);
+	reg_data &= I2S_CONF_LRCK_MASK;
+	return reg_data >> I2S_CONF_LRCK_FREQ_OFFSET;
+}
+
+/**
+ * @brief Enable i2s interrupt in event register
+ *
+ * @param reg base register of device
+ * @param irq_type irq type to be enabled one of I2S_EV_READY or I2S_EV_ERROR
+ *
+ * @return N/A
+ */
+static void i2s_irq_enable(int reg, int irq_type)
+{
+	assert(irq_type == I2S_EV_READY || irq_type == I2S_EV_ERROR);
+
+	uint8_t reg_data = litex_read8(reg + I2S_EV_ENABLE_REG_OFFSET);
+
+	litex_write8(reg_data | irq_type, reg + I2S_EV_ENABLE_REG_OFFSET);
+}
+
+/**
+ * @brief Disable i2s interrupt in event register
+ *
+ * @param reg base register of device
+ * @param irq_type irq type to be disabled one of I2S_EV_READY or I2S_EV_ERROR
+ *
+ * @return N/A
+ */
+static void i2s_irq_disable(int reg, int irq_type)
+{
+	assert(irq_type == I2S_EV_READY || irq_type == I2S_EV_ERROR);
+
+	uint8_t reg_data = litex_read8(reg + I2S_EV_ENABLE_REG_OFFSET);
+
+	litex_write8(reg_data & ~(irq_type), reg + I2S_EV_ENABLE_REG_OFFSET);
+}
+
+/**
+ * @brief Clear all pending irqs 
+ *
+ * @param reg base register of device 
+ *
+ * @return N/A
+ */
+static void i2s_clear_pending_irq(int reg)
+{
+	uint8_t reg_data = litex_read8(reg + I2S_EV_PENDING_REG_OFFSET);
+
+	sys_write8(reg_data, reg + I2S_EV_PENDING_REG_OFFSET);
```

Comment:
Why sys_write8 instead of litex_write8?

Reply:
For a single byte there is no difference between the two, but I agree that for consistency it should be `litex_write8`. It's fixed now.